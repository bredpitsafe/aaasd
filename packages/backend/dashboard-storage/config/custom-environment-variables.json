{
  "service": {
    "name": "SERVICE_NAME",
    "port": "SERVICE_PORT",
    "stage": "SERVICE_STAGE",
    "nodeNo": "SERVICE_NODE_NO",
    "url": "SERVICE_URL",
    "healthcheckURL": "SERVICE_HEALTHCHECK_URL"
  },
  "oauth": {
    "devUrl": "OAUTH_DEV_URL",
    "prodUrl": "OAUTH_URL"
  },
  "postgres": {
    "host": "POSTGRES_HOST",
    "port": "POSTGRES_PORT",
    "databaseName": "POSTGRES_DATABASE_NAME",
    "username": "POSTGRES_USERNAME",
    "password": "POSTGRES_PASSWORD",
    "minPoolSize": "POSTGRES_MIN_POOL_SIZE",
    "maxPoolSize": "POSTGRES_MAX_POOL_SIZE",
    "idleTimeout": "POSTGRES_IDLE_TIMEOUT",
    "connectionTimeout": "POSTGRES_CONNECTION_TIMEOUT",
    "allowExitOnIdle": "POSTGRES_ALLOW_EXIT_ON_IDLE",
    "scheme": "POSTGRES_SCHEME"
  },
  "metrics": {
    "url": "METRICS_URL",
    "prefix": "METRICS_PREFIX",
    "labels": "METRICS_LABELS"
  }
}
