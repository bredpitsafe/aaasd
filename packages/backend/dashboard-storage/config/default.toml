[service]
name = 'dashboard_storage'
host = '0.0.0.0'
port = 9090
httpServerPort = 9110
env = "dev"
httpHealthcheckURL = '/healthcheck'

[postgres]
host = 'localhost'
port = '5432'
databaseName = 'dashboards'
username = 'dashboards'
password = 'test'
minPoolSize = 1
maxPoolSize = 20
idleTimeout = 60_000
connectionTimeout = 2_000
allowExitOnIdle = true
maxConcurrentQueries = 1000
maxQueueSize = 10000
scheme = 'test'

[logging]
level = 'debug'

[metrics]
url = '/metrics'

[rpc]
timeout = 10000
