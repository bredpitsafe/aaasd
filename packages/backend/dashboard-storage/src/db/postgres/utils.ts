import { performance } from 'perf_hooks';
import type { Pool, QueryResult, QueryResultRow } from 'pg';

import { EMetricsLabel } from '../../def/metrics.ts';
import { metrics } from '../../utils/metrics.ts';
import { queryLimiter } from '../scheduler.ts';

export const operation = async <T extends QueryResultRow>(
    type: 'insert' | 'query' | 'delete',
    query: string,
    pool: Pool,
    metricsData: string,
): Promise<QueryResult<T>> => {
    const startTime = performance.now();
    try {
        const result = await queryLimiter.schedule(() => pool.query(query));
        metrics.db.totalQueries.inc({
            [EMetricsLabel.Type]: type,
            [EMetricsLabel.Value]: metricsData,
        });
        return result;
    } catch (err) {
        metrics.db.totalErroredQueries.inc({
            [EMetricsLabel.Type]: type,
            [EMetricsLabel.Value]: metricsData,
        });
        throw err;
    } finally {
        const duration = performance.now() - startTime;
        metrics.db.queryDuration.observe(
            { [EMetricsLabel.Type]: type, [EMetricsLabel.Value]: metricsData },
            duration,
        );
    }
};
