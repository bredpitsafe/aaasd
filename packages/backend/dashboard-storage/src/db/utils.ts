import { createHash } from 'crypto';
import { isEmpty } from 'lodash-es';

import type { TDbDashboard } from '../def/dashboard.ts';

export function getDigest(config: TDbDashboard['config']) {
    return isEmpty(config) ? '' : createHash('md5').update(config).digest('hex');
}
