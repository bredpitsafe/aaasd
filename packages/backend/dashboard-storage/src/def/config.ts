import type { EAppEnv } from '@common/types';

export type TConfig = {
    service: {
        name: string;
        host: string;
        port: number;
        httpServerPort: number;
        env: EAppEnv;
        httpHealthcheckURL: string;
    };
    rpc: {
        timeout: number;
    };
    postgres: {
        host: string;
        port: number;
        databaseName: string;
        username: string;
        password: string;
        minPoolSize: number;
        maxPoolSize: number;
        idleTimeout: number;
        connectionTimeout: number;
        allowExitOnIdle: boolean;
        maxConcurrentQueries: number;
        maxQueueSize: number;
        scheme: string;
    };
    metrics: {
        url: string;
        prefix: string;
        labels: [string, string][];
    };
    logging: {
        level: string;
    };
};
