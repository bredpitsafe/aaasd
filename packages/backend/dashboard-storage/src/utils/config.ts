import c from 'config';

import type { TConfig } from '../def/config.ts';

export const config = c as unknown as TConfig;
