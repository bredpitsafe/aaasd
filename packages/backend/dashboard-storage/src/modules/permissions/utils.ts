import { assertFail } from '@common/utils';
import type {
    TPermission,
    TSharePermission,
} from '@grpc-schemas/dashboard_storage-api-sdk/index.services.dashboard_storage.v1';

import type { TGrpcPermissionRow } from '../../def/permissions.ts';
import { EGroups } from '../../def/permissions.ts';

export function getMaxPermission<T extends Pick<TGrpcPermissionRow, 'permission' | 'user'>>(
    permissionRows: T[],
): { permissionRow: T; sharePermission: TSharePermission } {
    const getOrderIndex = (a: T): number =>
        (
            [
                'PERMISSION_OWNER',
                'PERMISSION_EDITOR',
                'PERMISSION_VIEWER',
                'PERMISSION_NONE',
            ] as TPermission[]
        ).findIndex((permission) => Number(a.permission === permission));
    const getUserIndex = (a: T): number => {
        if ((Object.values(EGroups) as string[]).includes(a.user)) {
            return 0;
        }
        return 1;
    };
    const sortedPermissionRows = [...permissionRows].sort((a, b) => {
        // Sort dashboards in permissions order, highest permission first.
        const orderDiff = getOrderIndex(a) - getOrderIndex(b);
        if (orderDiff !== 0) {
            return orderDiff > 0 ? 1 : -1;
        }
        // If some of the permissions have the same order, groups go first
        const userDiff = getUserIndex(a) - getUserIndex(b);

        return userDiff > 0 ? 1 : -1;
    });
    const [permissionRow] = sortedPermissionRows;
    const sharePermission: TSharePermission = permissionToSharePermission(
        sortedPermissionRows.find((p) => p.user === EGroups.All)?.permission ?? 'PERMISSION_NONE',
    );

    return {
        permissionRow,
        sharePermission,
    };
}

export function sharePermissionToPermission(sharePermission: TSharePermission): TPermission {
    switch (sharePermission) {
        case 'SHARE_PERMISSION_EDITOR':
            return 'PERMISSION_EDITOR';
        case 'SHARE_PERMISSION_VIEWER':
            return 'PERMISSION_VIEWER';
        case 'SHARE_PERMISSION_NONE':
            return 'PERMISSION_NONE';
        case 'SHARE_PERMISSION_UNSPECIFIED':
            assertFail(sharePermission);
    }
}

function permissionToSharePermission(permission: TPermission): TSharePermission {
    switch (permission) {
        case 'PERMISSION_OWNER':
        case 'PERMISSION_EDITOR':
            return 'SHARE_PERMISSION_EDITOR';
        case 'PERMISSION_VIEWER':
            return 'SHARE_PERMISSION_VIEWER';
        case 'PERMISSION_NONE':
            return 'SHARE_PERMISSION_NONE';
        case 'PERMISSION_UNSPECIFIED':
            assertFail(permission);
    }
}
