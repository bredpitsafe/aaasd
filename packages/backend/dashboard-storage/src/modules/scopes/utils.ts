import { GrpcResponseStatus } from '@backend/grpc/src/types/index.ts';
import { GrpcError } from '@backend/grpc/src/utils/error.ts';
import { isFlatObject } from '@common/utils';
import { isObject } from 'lodash-es';

const isScopeObject = (scope: unknown): boolean => {
    return isObject(scope) && isFlatObject(scope);
};

export const validateScopeObject = (scope: unknown) => {
    if (!isScopeObject(scope)) {
        throw new GrpcError({
            code: GrpcResponseStatus.INVALID_ARGUMENT,
            details:
                'Scope param must be a flat object with one level of primitive type properties',
        });
    }
};
