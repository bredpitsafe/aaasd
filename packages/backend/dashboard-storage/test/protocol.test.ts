import { generateTraceId } from '@common/utils';
import { wait } from '@common/utils/src/wait.ts';
import fetch from 'node-fetch';
import { firstValueFrom, fromEvent, fromEventPattern } from 'rxjs';
import { afterAll, describe, expect, it, vi } from 'vitest';
import WebSocket from 'ws';

import { config } from '../src/utils/config.js';
import { BASE_URL, Client, users, WS_ADDRESS } from './client/index.ts';

describe('Protocol', () => {
    const client = new Client();

    afterAll(() => {
        client.close();
    });

    it('healthcheck', async () => {
        const response = await fetch(`http://${BASE_URL}${config.service.httpHealthcheckURL}`);
        expect(await response.text()).toMatchSnapshot();
    });

    it('metrics', async () => {
        const response = await fetch(`http://${BASE_URL}${config.metrics.url}`);
        expect(response.status).toBe(200);
    });

    it('ping', async () => {
        const response = await client.request({ type: 'Ping' });
        expect(response.state).toBe('Done');
        expect(response.payload.type).toBe('Pong');
    });

    it('client heartbeat', async () => {
        const obs$ = client.send({ type: 'Heartbeat' });
        const subscriber = vi.fn();
        const subscription = obs$.subscribe(subscriber);
        await wait();

        expect(subscriber.mock.calls).toHaveLength(0);
        subscription.unsubscribe();
    });

    it('server heartbeat', async () => {
        const obs$ = client.send({ type: 'ServerHeartbeat' });
        const subscriber = vi.fn();
        const subscription = obs$.subscribe(subscriber);
        await wait(2000);

        expect(subscriber.mock.calls.length).toBeGreaterThan(0);
        expect(subscriber.mock.calls[0][0]).toMatchObject({
            state: 'InProgress',
            payload: {
                type: 'Heartbeat',
            },
        });
        subscription.unsubscribe();
    });

    it('traceId', async () => {
        const traceId = generateTraceId();
        const response = await client.request({ type: 'Ping' }, { traceId });
        expect(response.traceId).toBe(traceId);
    });

    it('correlationId', async () => {
        const correlationId = Math.floor(Math.random() * 1000);
        const response = await client.request({ type: 'Ping' }, { correlationId });
        expect(response.correlationId).toBe(correlationId);
    });

    it('unknown message type', async () => {
        const response = await client.request({ type: 'UNKNOWN_MESSAGE' });
        expect(response.error).toMatchSnapshot();
    });

    it('missing traceId', async () => {
        const response = await client.request({ type: 'UNKNOWN_MESSAGE' }, { traceId: undefined });
        expect(response.error).toMatchSnapshot();
    });

    it('missing correlationId', async () => {
        const response = await client.request({ type: 'UNKNOWN_MESSAGE' }, { traceId: undefined });
        expect(response.error).toMatchSnapshot();
    });

    it('non-JSON payload', async () => {
        const ws = new WebSocket(WS_ADDRESS);
        await firstValueFrom(fromEvent(ws, 'open'));
        const handleMessage = (handler: Function) =>
            ws.on('message', (message) => {
                const response = JSON.parse(message.toString());
                handler(response);
            });
        const promise = firstValueFrom<{ state: string; error: object }>(
            fromEventPattern(handleMessage),
        );
        ws.send('NOT_A_JSON');
        const message = await promise;
        expect(message.state).toBe('Done');
        expect(message.error).toMatchSnapshot();
    });

    it('simulate internal error', async () => {
        const response = await client.request({ type: 'Ping', simulateInternalError: true });

        expect(response.error).toMatchSnapshot();
    });

    it('simulate subscription timeout', async () => {
        const authenticatedClient = new Client({ authenticate: true, as: users.user1.username });
        const response = await authenticatedClient.request({ type: 'Ping', simulateTimeout: true });

        expect(response.error).toMatchSnapshot();
    }, 10_000);
});
