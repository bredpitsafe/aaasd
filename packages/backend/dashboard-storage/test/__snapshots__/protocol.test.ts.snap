// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Protocol > healthcheck 1`] = `"Ok."`;

exports[`Protocol > missing correlationId 1`] = `
{
  "description": "Request validation error: request body is missing required fields since it does not match any request schema.",
  "kind": "Validation",
}
`;

exports[`Protocol > missing traceId 1`] = `
{
  "description": "Request validation error: request body is missing required fields since it does not match any request schema.",
  "kind": "Validation",
}
`;

exports[`Protocol > non-JSON payload 1`] = `
{
  "args": {
    "originalError": "Unexpected token 'N', "NOT_A_JSON" is not valid JSON",
  },
  "description": "Request processing error. internal error occurred while processing the request. Please file a bug report and provide client logs.",
  "kind": "InternalError",
}
`;

exports[`Protocol > simulate internal error 1`] = `
{
  "description": "Non-actor error: Simulated Internal Error",
  "kind": "Unspecified",
}
`;

exports[`Protocol > simulate subscription timeout 1`] = `
{
  "description": "Subscription timeout: Subscription failed to produce first value after 5000ms",
  "kind": "Timeout",
}
`;

exports[`Protocol > unknown message type 1`] = `
{
  "description": "Request validation error: request body is missing required fields since it does not match any request schema.",
  "kind": "Validation",
}
`;
